File: backend/schemas/portfolio.py
from pydantic import BaseModel, Field, field_validator


class PortfolioHolding(BaseModel):
    symbol: str
    qty: float
    avg_price: float
    curr_price: float
    pnl: float


class PortfolioSummary(BaseModel):
    total_value: float
    holdings: list[PortfolioHolding] = Field(default_factory=list)


class PortfolioHoldingCreate(BaseModel):
    ticker: str
    qty: float
    price: float

    @field_validator("ticker")
    @classmethod
    def normalize_ticker(cls, value: str) -> str:
        normalized = value.strip().upper()
        if not normalized:
            raise ValueError("Ticker is required.")
        return normalized

    @field_validator("qty", "price")
    @classmethod
    def positive_number(cls, value: float) -> float:
        if value <= 0:
            raise ValueError("Value must be greater than zero.")
        return value

File: src/types/portfolio.ts
export interface PortfolioHolding {
  symbol: string
  qty: number
  avg_price: number
  curr_price: number
  pnl: number
}

export interface PortfolioSummary {
  total_value: number
  holdings: PortfolioHolding[]
}

export interface PortfolioHoldingInput {
  ticker: string
  qty: string
  price: string
}
